tools/loader/ModelCompiler.cpp

```diff
@@ -0,0 +1,186 @@
+        "   - bool (logic type)\n"),
+    llvm::cl::value_desc("name,type,shape"), llvm::cl::cat(modelCompilerCat));
+
+} // namespace
+
+/// Parse the strings and get the model input names and types.
+static void getModelInputs(std::vector<std::string> &inputNames,
+                           std::vector<Type> &inputTypes) {
+  for (const auto &str : modelInputs) {
+    // Split string. Expected format:
+    // - <name>,<type>,<shape> for non-quantized type.
+    // - <name>,<type>,<scale>,<offset>,<shape> for quantized type.
+    auto strPair = llvm::StringRef(str).split(',');
+    llvm::StringRef name = strPair.first;
+    strPair = strPair.second.split(',');
+    llvm::StringRef type = strPair.first;
+    CHECK(name.size()) << "Model input name empty";
+    CHECK(type.size()) << "Model input type empty";
+
+    // Parse type string.
+    ElemKind kind;
+    if (type.equals("float")) {
+      kind = ElemKind::FloatTy;
+    } else if (type.equals("float16")) {
+      kind = ElemKind::Float16Ty;
+    } else if (type.equals("int8q")) {
+      kind = ElemKind::Int8QTy;
+    } else if (type.equals("int16q")) {
+      kind = ElemKind::Int16QTy;
+    } else if (type.equals("int32q")) {
+      kind = ElemKind::Int32QTy;
+    } else if (type.equals("int32")) {
+      kind = ElemKind::Int32ITy;
+    } else if (type.equals("int64")) {
+      kind = ElemKind::Int64ITy;
+    } else if (type.equals("bool")) {
+      kind = ElemKind::BoolTy;
+    } else {
+      LOG(FATAL) << strFormat("Model input type %s not supported",
+                              std::string(type).c_str());
+    }
+
+    // For quantized type get scale and offset.
+    double scale;
+    int32_t offset;
+    if (isQuantizedElemKind(kind)) {
+      strPair = strPair.second.split(',');
+      CHECK(strPair.first.size()) << "Model input scale empty";
+      CHECK(!strPair.first.getAsDouble(scale))
+          << "Model input scale parameter invalid";
+      strPair = strPair.second.split(',');
+      CHECK(strPair.first.size()) << "Model input offset empty";
+      CHECK(!strPair.first.getAsInteger(0, offset))
+          << "Model input offset parameter invalid";
+    }
+
+    // Parse shape string.
+    llvm::StringRef shape = strPair.second;
+    CHECK(shape.size()) << "Model input shape empty";
+    ShapeVector dims;
+    CHECK_EQ(shape.front(), '[') << "First shape char should be [";
+    shape = shape.drop_front();
+    CHECK_EQ(shape.back(), ']') << "First shape char should be ]";
+    shape = shape.drop_back();
+    CHECK(shape.size()) << "Model input shape empty";
+    size_t val;
+    while (shape.contains(',')) {
+      auto splitRes = shape.split(',');
+      CHECK(!splitRes.first.getAsInteger(0, val))
+          << "Model input shape integer invalid";
+      dims.push_back(val);
+      shape = splitRes.second;
+    }
+    CHECK(!shape.getAsInteger(0, val)) << "Model input shape integer invalid";
+    dims.push_back(val);
+
+    // Push data.
```

Comment:
Do we need to check if there are multiple inputs with the same name?

Reply:
I`m not sure. It would be a weird (ill defined) graph in my opinion to have multiple placeholders with same name.

Reply:
Exactly! But a user could specify the same input multiple times on the command line. I'd say we should detect such cases here and bail with an error message.

Reply:
I moved the utility in the Loader class and added a checker for unique name. Have a look.